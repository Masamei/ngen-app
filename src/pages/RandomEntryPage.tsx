import { useEffect, useState, memo } from 'react'
import EntryCard from '../components/EntryCard'
import '../output.css'
import { Entry } from '../interfaces'
import mockData from '../assets/random.json'
import fetchData from '../services/fetchService'

function RandomEntryPage() 
{
  const [entry, setEntry] = useState<Entry | null>(null)
  const [loading, setLoading] = useState<boolean>(true)
  // const [error, setError] = useState<string | null>(null)

  async function fetchAndSetData() {
    setLoading(true)
    await fetchData<{entries: Entry[], error: string}>('https://api.publicapis.org/random')
    // const result = await fetchData<{entries: Entry[], error: string}>('https://api.publicapis.org/random')
    // setEntry(result.response?.entries[0] ?? null)
    // setError(result.error)
    setLoading(false)
  }

  useEffect(() => {
    fetchAndSetData()
    //just for testing purposes, because api is down
    setEntry(mockData.entries[0])
  }, [])

  if (loading) return <div>Loading...</div>
  
  return (
    <div className='flex flex-col items-between'>
      
      {/* {error && <div>{error}</div>} */}
      {/* {entry && !error && <EntryCard entry={entry} />} */}
      {entry && <EntryCard entry={entry} />}
      <button onClick={fetchAndSetData} className='mt-4 p-2 bg-blue-500 rounded'>Get different random entry</button>
    </div>
  )
}

export default memo(RandomEntryPage)