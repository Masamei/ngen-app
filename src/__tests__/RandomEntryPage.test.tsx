import { render, screen, waitFor, act } from '@testing-library/react'
import RandomEntryPage from '../pages/RandomEntryPage'

const mockFetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve({
      entries: [
        {
          API: 'The Muse',
          Description: 'Job board and company profiles',
          Auth: 'apiKey',
          HTTPS: true,
          Cors: 'unknown',
          Link: 'https://www.themuse.com/developers/api/v2',
          Category: 'Jobs',
        },
      ],
    }),
  })
)

global.fetch = mockFetch as jest.Mock

describe('RandomEntryPage Component', () => {
  beforeEach(() => {
    mockFetch.mockClear()
  })

  test('fetches and displays a random entry', async () => {
    await act(async () => {
      render(<RandomEntryPage />)
    })

    await waitFor(() => {
      expect(screen.getByText('The Muse')).toBeInTheDocument()
      expect(screen.getByText('Job board and company profiles')).toBeInTheDocument()
    })

    expect(mockFetch).toHaveBeenCalledTimes(1)
  })

  test.skip('displays error message on fetch failure', async () => {
    mockFetch.mockImplementationOnce(() => Promise.reject('API is down'))

    await act(async () => {
      render(<RandomEntryPage />)
    })

    await waitFor(() => {
      expect(screen.getByText('API is down')).toBeInTheDocument()
    })
  })

  test('matches snapshot', async () => {
    const { asFragment } = render(<RandomEntryPage />)

    await waitFor(() => {
      expect(screen.getByText('The Muse')).toBeInTheDocument()
    });

    expect(asFragment()).toMatchSnapshot()
  })
})