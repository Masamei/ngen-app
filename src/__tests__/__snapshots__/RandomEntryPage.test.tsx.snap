// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RandomEntryPage Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col items-between"
  >
    <div
      class="flex flex-wrap justify-evenly"
    >
      <div
        class="cursor-pointer text-silver bg-customBackground hover:bg-silver hover:text-green  rounded-lg p-4 m-2 min-w-52 max-w-52 w-52 transition duration-300 ease-in-out"
      >
        <h2
          class="text-lg font-bold mt-2 elipsis-pretty"
        >
          The Muse
        </h2>
        <p
          class="text-gray-400 mb-2 text-[8px]"
        >
          Category: Jobs
        </p>
        <p
          class="text-gray-400 pb-4 text-xs"
        >
          Job board and company profiles
        </p>
        <span
          class="grid grid-cols-5 gap-1"
        >
          <p
            class="text-gray-400 pb-4 text-[8px] text-green"
          >
            HTTPS
          </p>
          <p
            class="text-gray-400 pb-4 text-[8px] col-span-2"
          >
            Cors: unknown
          </p>
          <p
            class="text-gray-400 pb-4 text-[8px] col-span-2"
          >
            AUTH: apiKey
          </p>
        </span>
      </div>
    </div>
    <button
      class="mt-4 p-2 bg-blue-500 rounded"
    >
      Get different random entry
    </button>
  </div>
</DocumentFragment>
`;
